### DISCUSSION TITLE
Custom metrics

### DISCUSSION THREAD

**Discussion by janiceylau** [talmolab/sleap#1175] [General]
Hello! Is it possible to use custom metrics for evaluation, as well as custom losses for training?

**Comment by roomrys** [comment]
Hi @janiceylau,

When training is run, SLEAP calls the `Training.evaluate()` function which calls the `slean/nn/evals.py/evaluate_model()` on each split of the labels (training, validation, and test). 
https://github.com/talmolab/sleap/blob/05683d51f2da34d4a684d962e9765cb57dfd87c9/sleap/nn/evals.py#L650-L738

The `evaluate_model()` function then runs prediction and calls `slean/nn/evals.py/evaluate` to evaluate how good that prediction is based on SLEAP's metrics. That being said...

All of SLEAP's evaluation is done by the `slean/nn/evals.py/evaluate` function:
https://github.com/talmolab/sleap/blob/05683d51f2da34d4a684d962e9765cb57dfd87c9/sleap/nn/evals.py#L585-L647

If you were looking for a place to add your own custom metric, this would be the function you are looking for. Do you already have [SLEAP installed from source](https://sleap.ai/installation.html#conda-from-source)? Then you can readily edit this function to satisfy your needs.

Let me know if you need any help and if this was (or was not) what you were looking for.

Thanks,
Liezl

**Comment by rustamdantia** [comment]
@roomrys
Hi Liezl,

Thank you for your answer :)  Is there not a convenient way of adding custom metrics _without_ modifying the SLEAP source code ? From a software engineering perspective, e.g. when writing an application that depends on SLEAP, that would be desirable. One possibility is just copying and modifying the `evaluate` function in [sleap/sleap/nn/evals.py](https://github.com/talmolab/sleap/blob/05683d51f2da34d4a684d962e9765cb57dfd87c9/sleap/nn/evals.py#L585-L647) as well as all the code that calls it (e.g. the training code) and replacing all the calls to the original evalute function with calls to the modified version. While that avoids modifying the SLEAP source code it confusingly creates a lot of unnecessary irrelevant duplicated code that should be abstracted away. I'm wondering if there is a better way.

Thanks, 

Rustam
